### DISCUSSION TITLE
git commit doesn't add to current branch

### DISCUSSION THREAD

**Discussion by bcspragu** [arxanas/git-branchless#520] [Q&A]
I've got a setup that looks something like:

```
◇ 67d7ee3 7h (main, remote origin/main) [...]
┃
◯ 0ccfe18 6h Commit #1
┃
◯ e5e6e68 2h Commit #2
┃
◯ 72a4e46 2h Commit #3
┃
◯ 60a10d8 1h (brandon/first-branch) Commit #4, out for review in PR1
┃
● 981808f 22m (brandon/second-branch) Commit #5, still working on it
┃
◯ 30e9a32 4m Commit #6, which I thought would get updated as brandon/second-branch, but didn't
```

I don't think I did anything special here, just:

```bash
git co # select brandon/second-branch
vim ... # write some code
git commit # commit the new stuff, 'Commit #6' in this example
```

I'd expect the branch pointer for `brandon/second-branch` to be on `Commit #6`, which is what usually happens, but it looks like I'm in a detatched HEAD state now?

My questions:
1. How did I end up here?
2. How do I fix it (e.g. move the branch pointer from `981808f` to `30e9a32`) without breaking everything?

**Comment by bcspragu** [comment]
I was able to fix it with `git branch -f brandon/second-branch 30e9a32`, but still not sure how I got into that state.

As a test, I checkout out `brandon/second-branch` and then ran:

```bash
git co -b brandon/second-branch-2
vim ... # write some code
git commit
```

And it worked as expected, `brandon/second-branch-2` was updated to point to the latest commit

**Reply by bcspragu** [reply]
Oh, it seems like `git co` is not actually the same as `git checkout`, seems like I need to learn how to use this tool better.

**Reply by claytonrcarter** [reply]
Hi there. I'm not very familiar w/ the interactive mode of `git co`, but it looks like it leaves you in a detached HEAD by design. So even though you selected the commit w/ your intended branch name, it just checked out *that commit*, not the branch. You can also use `git switch <branch>` or `git checkout <branch>` to checkout a branch. Your example with `git co -b brandon/second-branch-2` worked as expected because specifying `-b` will bypass the interactive mode and act just like `git checkout -b <branch>`.

There are several other options for moving branches around. One is to just delete it and recreate it: `git branch -D <branch>; git branch <branch> <commit>`. Another is move it manually: `git branch -f <branch> <commit>`. Note that the latter doesn't work if you have that branch checked out.

**Reply by bcspragu** [reply]
Ah, I guess that's the distinction, I usually specify the branch name directly, I just thought the interactive mode would be more convenient. Good to know, thanks!

**Reply by arxanas** [reply]
@bcspragu Also check out the thread at https://github.com/arxanas/git-branchless/discussions/402 for some ideas on how to manipulate the branch pointer. I added a link to that thread to the Wiki.

For `git co`, it usually does delegate to `git checkout`, but like @claytonrcarter mentioned, it doesn't check out branches when using the interactive mode. I think there's a good case to be made for git-branchless to automatically check out the associated branch when moving to a given commit, whether by `git co` or by using `git next`/`git prev`. If you're interested in implementing that, I'm happy to guide you 🙂.

**Reply by bcspragu** [reply]
I'm definitely interested in trying to implement that, any pointers would be supremely useful

**Reply by bcspragu** [reply]
Sent out https://github.com/arxanas/git-branchless/pull/526 as a first attempt, let me know what you think
